compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java

```diff
@@ -3815,7 +3824,7 @@ private ParserRuleContext getNextRuleForKeywords(ParserRuleContext currentCtx, i
                 return ParserRuleContext.EXPRESSION;
             case FAIL_KEYWORD:
                 if (getParentContext() == ParserRuleContext.ON_FAIL_CLAUSE) {
-                    return ParserRuleContext.TYPE_DESC_IN_TYPE_BINDING_PATTERN;
+                    return ParserRuleContext.OPTIONAL_TYPE_DESC_IN_TYPE_BINDING_PATTERN;
```

Comment:
Error handler changes look good.

Shall we rename the `OPTIONAL_TYPE_DESC_IN_TYPE_BINDING_PATTERN` to something like `ON_FAIL_OPTIONAL_BINDING_PATTERN` considering,
1. the whole binding pattern is optional, not only the type descriptor
2. optional binding pattern is specific to on-fail-clause